### ISSUE TITLE
"every possible result is equally likely"

### ISSUE THREAD

**Issue by igel-kun** [ciphergoth/sansreplace#2]
I stumbled upon your code via [Stackoverflow](https://stackoverflow.com/a/67850443/6470423), and [I made a comment there](https://stackoverflow.com/questions/311703/algorithm-for-sampling-without-replacement/46807110#comment138161224_67850443) because I'm not convinced that all outputs are equally likely. I'll repeat the comment here for completeness:

Suppose n is large and consider all possible outcomes of the algorithm. Such an outcome can be characterized by the array R of numbers returned by [the random number generator function]. If I understood correctly, [1,3,..,2k-1] will be output at least k! times (since R being any permutation of [1,2,...,k] will lead to this), but [0,1,...,k-1] is returned only twice: if R=[0,0,...,0] or if R=[t,t,...,t].

But I might be wrong, do you happen to have any proof that all k-sets are in fact equally likely?

**Comment by igel-kun** [comment]
Maybe statistical evidence would be good enough, letting the code select 4 out of 12 numbers a billion times and then visualizing the histogram.

**Comment by ciphergoth** [comment]
Considering the 12-choose-4 case as you suggest, you are correct that any permutation of [0, 1, 2, 3] will lead to the output [0, 2, 4, 6], while only [0, 0, 0, 0] leads to [0, 1, 2, 3]. However, we compensate for this when generating R by building in a bias towards re-selecting what was selected earlier. When R = [0] and we are selecting the next element, 0 is twice as likely as 1; when R = [0, 0], 0 is three times as likely as 1. Overall [0, 0, 0, 0] is 4! times more likely than [0, 1, 2, 3], exactly compensating for the bias you observe.

Have you read https://github.com/ciphergoth/sansreplace/blob/master/cardchoose.md ? That should explain why this is unbiased.

Thanks for commenting!

**Comment by ciphergoth** [comment]
Whoops used the wrong closing status sorry!

**Comment by ciphergoth** [comment]
Whoops used the wrong closing status sorry!

**Comment by igel-kun** [comment]
Right, that way of thinking about it makes alot of sense, thanks for the pointer.
So, from the perspective of what the RNG returns, it's not only [0,0,0,0] that will come out as [0,1,2,3], but also [0,9,10,11] and [0,9,9,0] and so forth. Indeed, it turns out that [0,1,2,3] is returned for exactly k! of all n!/(n-k)! possible arrays returned by the RNG, namely everyone in {0} x {0,t} x {0,t,t+1} x  ... x {0,t,t+1,...,t+k-1}, which are  1 * 2 * 3 * .. * k many. Nice.

**Comment by ciphergoth** [comment]
Yes (assuming this is the 4-choose-9 case) - that's a better way to put it than my comment, thanks!

**Comment by igel-kun** [comment]
Hum? I thought it was the 4-choose-12 case. If the RNG draws [0,0,0,8] it will get translated to [0,1,2,11] no? 11 should not be drawable in 4-choose-9. Maybe I have a different understanding of what "k-choose-n" means :)

Anyways, I ran some benchmarks and, on my machine, it seems that the Fisher-Yates adaptation mentioned in the Stackoverflow has Cardchoose beat for "larger" k, starting at about k = 60. See my benchmark code below, compiled with `g++ -std=c++20 -O3` but here's the output on my machine (n=1000 but it shouldn't matter):
```
k = 20	card:	19.4ms	fiya:	24.8ms
k = 25	card:	25.9ms	fiya:	32.4ms
k = 30	card:	32.7ms	fiya:	35.3ms
k = 35	card:	39.6ms	fiya:	41.9ms
k = 40	card:	46.7ms	fiya:	48.1ms
k = 45	card:	54.0ms	fiya:	55.7ms
k = 50	card:	61.0ms	fiya:	62.8ms
k = 55	card:	68.5ms	fiya:	70.2ms
k = 60	card:	76.0ms	fiya:	68.7ms
k = 65	card:	83.5ms	fiya:	74.4ms
k = 70	card:	91.3ms	fiya:	81.3ms
k = 75	card:	98.8ms	fiya:	87.2ms
k = 80	card:	106.5ms fiya:	93.6ms
k = 85	card:	114.6ms fiya:	100.3ms
```

```C++
#include <vector>
#include <unordered_map>
#include <cstdint>
#include <ranges>
#include <algorithm>
#include <chrono>
#include <iostream>
#include <iomanip>

using namespace std::chrono;
using Point = steady_clock::time_point;
using Duration = steady_clock::duration;

Point get_time() { return steady_clock::now(); }
double ms_between(const Point& t1, const Point& t2) { return duration<double, std::milli>(t2-t1).count(); }

template<class Call, class... Args> requires (std::is_void_v<std::invoke_result_t<Call, Args&&...>>)
double timed_call(Call&& callable, Args&&... args) {
  auto p = get_time();
  callable(std::forward<Args>(args)...);
  auto q = get_time();
  return ms_between(p, q);
}

//! return the result of throwing a die with 'sides' sides [0,sides-1]
inline uint32_t throw_die(const uint32_t sides = 6) {
  return rand() % sides; // yadda yadda, it's not 100% uniform for tiny values of RAND_MAX...
}

std::unordered_map<uint32_t, uint32_t> track;
template<class Set = std::vector<uint32_t>>
void fisher_yates_choose(uint32_t k, uint32_t n, Set& result) {
  // we use a variant of Fisher-Yates shuffle, tracking only the changed elements of [0,..,n-1]
  // (see https://stackoverflow.com/questions/196017/unique-non-repeating-random-numbers-in-o1#comment138160976_44884435)
  // entry track[x] = y means that we've replaced x with y in an iteration before
  track.clear();
  for (uint32_t i = 0; i != k; ++i) result.push_back(i);
  for (uint32_t i = 0; i != k; ++i) {
    const uint32_t r = throw_die(n--) + i;

    if(r >= k) {
      const auto res_i = result[i];
      const auto [fr, success] = track.emplace(r, res_i);
      if(!success) {
        result[i] = fr->second;
        fr->second = res_i;
      } else result[i] = r;
    } else std::swap(result[i], result[r]);
  }
}

// probs to Paul Crowley for this one: https://stackoverflow.com/questions/311703/algorithm-for-sampling-without-replacement/67850443#67850443
template<class Set = std::vector<uint32_t>>
void cardchoose(const uint32_t k, const uint32_t n, Set& result) {
  const uint32_t t = n - k + 1;
  for(uint32_t i = 0; i != k; ++i) {
    const uint32_t r = throw_die(t + i);
    result.push_back((r < t) ? r : result[r - t]);
  }
  std::ranges::sort(result);
  for(uint32_t i = 0; i != k; ++i)
    result[i] += i;
}

template<int algo>
void test(const uint32_t k, const uint32_t n, uint32_t num_runs) {
  std::vector<uint32_t> result;
  result.reserve(k);
  while(num_runs--) {
    result.clear();
    if constexpr (algo == 0)
      cardchoose(k, n, result);
    else if constexpr (algo == 1)
      fisher_yates_choose(k, n, result);
  }
}

int main() {
  std::cout << std::setprecision(1) << std::fixed;
  const uint32_t n = 1000;
  for(uint32_t k = 20; k < 90; k += 5){
    std::cout << "k = "<<k
              << "\tcard:\t" << timed_call(test<0>, k, n, 30000) << "ms"
              << "\tfiya:\t" << timed_call(test<1>, k, n, 30000) << "ms\n";
  }
}
```
Since the Fischer-Yates adaptation is O(k) and Cardchoose is O(k log k) it kinda makes sense that one overtakes the other at some point, depending on k. Of course, that point depends on the machine, glibc, etc etc

**Comment by ciphergoth** [comment]
I'd be interested to know what results you get from running the benchmarking code in this repository? See also the graphs I've shared here: https://github.com/ciphergoth/sansreplace/blob/master/results.md

**Comment by igel-kun** [comment]
Hmm, I'm not good enuff in Python to understand how tests are being conducted in the benchmark here, but I noticed that the Ficher-Yates adaptation was already in the comparison under the name `hsel`, so I put my improvements into the existing code there (see #3).
Looking at the implementation, I noticed that the code I gave above is faulty and I'll repair it in a minute. The code in the pull-request I made should be correct and contains the following improvements:
1. I made the hash-map that's supposed to track changes to the "virtual [0,n-1]-array" global, so I'm not paying construction+destruction each time, just `clear()` which is much cheaper.
2. I'm representing the lower k cells of the "virtual [0,n-1]-array" in `result` itself, so I don't have to look-up in the hash-map whether the current i has been replaced in the past, but I can look that up in the `result` array, which is faster since it's an array and the memory access `result[i]` has to be cached anyways
